CRM/Financial/BAO/FinancialItem.php

```diff
@@ -286,25 +286,23 @@ public static function checkContactPresent($contactIds, &$error) {
   }
 
   /**
-   * Get last financial item data.
+   * Get non-liability financial item with the highest id relating to the line item.
    *
-   * @param int $entityId
+   * In practical terms this means return the row that is not associated with
+   * sales tax as sales tax is a liability account.
    *
-   * @param string $entityTable
+   * @param int $entityId
    *
    * @return object CRM_Core_DAO
    */
-  public static function getPreviousFinancialItem($entityId, $entityTable = 'civicrm_line_item') {
-    $queryParams = array(
-      1 => array($entityId, 'Integer'),
-      2 => array($entityTable, 'String'),
-    );
-    $query = 'SELECT id, description, status_id, financial_account_id 
-      FROM civicrm_financial_item
-      WHERE entity_id = %1 AND entity_table = %2 ORDER BY id DESC LIMIT 1';
-    $prevFinancialItem = CRM_Core_DAO::executeQuery($query, $queryParams);
-    $prevFinancialItem->fetch();
-    return $prevFinancialItem;
+  public static function getPreviousNonLiabilityFinancialItem($entityId) {
+    $liabilityFinancialAccounts = civicrm_api3('FinancialAccount', 'get', array('financial_account_type_id' => 'Liability'));
+    return civicrm_api3('FinancialItem', 'getsingle', array(
+      'entity_id' => $entityId,
+      'entity_table' => 'civicrm_line_item',
+      'financial_account_id' => array('NOT IN' => array_keys($liabilityFinancialAccounts['values'])),
```

Comment:
For Deferred revenue entries financial_account_id is of Liability financial account type, so i feel this api will return 0 records therefore i had used  query to get details of last financial item.

```
 SELECT cfi.id, cfi.description, cfi.amount, cfi.financial_account_id, cfi.status_id
      FROM `civicrm_financial_item` cfi
        INNER JOIN civicrm_entity_financial_trxn ceft 
          ON ceft.entity_id = cfi.id AND ceft.entity_table = "civicrm_financial_item"
        INNER JOIN civicrm_financial_trxn cft ON cft.id = ceft.financial_trxn_id
        INNER JOIN civicrm_financial_account cfa ON cfa.id = cfi.financial_account_id
      WHERE cfi.entity_table ='civicrm_line_item' AND cfi.entity_id =2 AND 
      ((cfa.financial_account_type_id = 2 AND cfi.financial_account_id = from_financial_account_id) OR financial_account_type_id <> 2)
      ORDER BY cfi.id DESC LIMIT 1
```